### PR TITLE
Typescript: Fix types of client-api & storystore

### PR DESCRIPTION
**Pull request by ndelangen**
## What I did
- the function to merge all decorators & storyFn has been changed to no longer mutate
- the storyFn called from within decorators must be called with a StoryContext
- add a story for vue to test the behaviour

### CODE REVIEW

**Review by shilman — APPROVED** [review]
LGTM let's see what @tmeasday @igor-dv @Hypnosphi say about this hopefully minor breaking change

**Comment by ndelangen** [comment]
I'd love to merge this ASAP, so I can continue with the app/react & app/angular mergers

File: lib/client-api/src/story_store.ts [hunk 1]
```diff
@@ -188,8 +186,12 @@ export default class StoryStore extends EventEmitter {
       applyDecorators(getOriginal(), getDecorators())
     );
 
-    const storyFn: StoryFn = (p: object) =>
-      getDecorated()({ ...identification, parameters: { ...parameters, ...p } });
+    const storyFn: StoryFn = p =>
+      getDecorated()({
+        ...identification,
+        ...p,
+        parameters: { ...parameters, ...(p && p.parameters) },
```
Comment (Jessica-Koch) [inline lib/client-api/src/story_store.ts]: i wasn't aware you could do `...(p && p.parameters)`, learned something new

File: examples/vue-kitchen-sink/src/stories/custom-decorators.stories.js [hunk 2]
```diff
@@ -30,6 +30,11 @@ export const template = () => ({
   template: '<my-button>MyButton with template</my-button>',
 });
 
+export const withData = ({ parameters, ...rest }) =>
+  console.log(parameters) || {
+    template: `<pre>${JSON.stringify({ ...rest, parameters }, null, 2)}</pre>`,
+  };
```
Comment (Jessica-Koch) [inline examples/vue-kitchen-sink/src/stories/custom-decorators.stories.js]: you dont' want the console to come after the other template string?  I'm not sure what the main goal of this function is trying to achieve via the naming of it.

**Review by Jessica-Koch — APPROVED** [review]

**Comment by ndelangen** [comment]
ping @Hypnosphi @igor-dv we'd really like your opinion about this change.

File: app/vue/src/client/preview/index.js [hunk 3]
```diff
@@ -46,13 +46,18 @@ function decorateStory(getStory, decorators) {
 
       const decoratedStory = decorator((p = {}) => {
         story = decorated(
-          Object.assign(
-            context,
-            p,
-            { parameters: Object.assign(context.parameters || {}, p.parameters) },
-            { options: Object.assign(context.options || {}, p.options) }
```
Comment (Hypnosphi) [inline app/vue/src/client/preview/index.js]: What happens with `options`?

Comment (shilman) [reply app/vue/src/client/preview/index.js]: Options should never have been there in the first place AFAIK, so we deleted it. Technically a breaking change, but I think this was a bug & all code that I know about would use `parameters.options` which is accounted for in this change.

**Review by Hypnosphi — APPROVED** [review]

File: lib/api/src/version.ts [hunk 4]
```diff
@@ -1 +1 @@
-export const version = '5.2.0-alpha.39';
```
Comment (shilman) [inline lib/api/src/version.ts]: @ndelangen Please don't modify this file. It only causes merge conflicts down the line.

**Pull request merged**